subprojects/core/src/main/java/org/gradle/api/internal/project/DefaultProjectStateRegistry.java

```diff
@@ -196,5 +197,49 @@ public T create() {
         public boolean hasMutableState() {
             return workerLeaseService.getCurrentProjectLocks().contains(projectLock);
         }
+
+        @Override
+        public SafeExclusiveLock newSafeExclusiveLock() {
```

Comment:
Safe is a little bit of a weird label for this.  I think you mean "safe when used with project state locks", but maybe we should consider something like... `newExclusiveOperationLock` since it's intended to be a safe way to exclusively do an operation where the project lock may be held/released?